k8s/forward/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+//! kubernetes api-server.
+//!
+//! If you're looking at a higher-level construct, please take a look at kube-proxy.
+
+mod http_forward;
+mod pod_selection;
+mod port_forward;
+
+/// Layer 7 proxies.
+pub use http_forward::{HttpForward, HttpProxy};
+/// Layer 4 proxies.
+pub use port_forward::PortForward;
+
+/// The kubernetes api version used throughout the crate.
+pub(crate) use k8s_openapi::api::core::v1 as vx;
+
+use anyhow::Context;
+use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
+use kube::ResourceExt;
+use vx::Pod;
+
+/// Different types of target selectors.
+#[derive(Clone)]
+pub enum TargetSelector {
+    /// By pod name.
+    PodName(String),
+    /// By pod label selector.
+    PodLabel(String),
+    /// By service label selector.
+    ServiceLabel(String),
+}
+impl TargetSelector {
+    /// New `Self` from the given pod label key value.
+    pub fn pod_label(key: &str, val: &str) -> Self {
+        Self::PodLabel(format!("{}={}", key, val))
+    }
+    /// New `Self` from the given service label key value.
+    pub fn svc_label(key: &str, val: &str) -> Self {
+        Self::ServiceLabel(format!("{}={}", key, val))
+    }
+}
+
+/// Identify a port explicitly by its number of by name.
+#[derive(Clone)]
+pub enum Port {
+    /// Specified using a number.
+    Number(i32),
+    /// Specified using a name.
+    Name(String),
+}
+impl From<i32> for Port {
+    fn from(port: i32) -> Self {
+        Self::Number(port)
+    }
+}
+impl From<&str> for Port {
+    fn from(port: &str) -> Self {
+        Self::Name(port.to_string())
+    }
+}
+impl From<IntOrString> for Port {
+    fn from(port: IntOrString) -> Self {
+        match port {
+            IntOrString::Int(port) => Self::Number(port),
+            IntOrString::String(port) => Self::Name(port),
+        }
+    }
+}
+impl Port {
+    /// Returns the port name, if set.
+    pub(crate) fn name(&self) -> Option<&String> {
+        match self {
+            Port::Number(_) => None,
+            Port::Name(name) => Some(name),
+        }
+    }
+    /// Returns the port number, if set.
+    pub(crate) fn number(&self) -> Option<i32> {
+        match self {
+            Port::Number(number) => Some(*number),
+            Port::Name(_) => None,
+        }
+    }
+    /// Returns the port as a string.
+    pub(crate) fn any(&self) -> String {
+        match self {
+            Port::Number(number) => number.to_string(),
+            Port::Name(name) => name.clone(),
+        }
+    }
+}
+
+/// A kubernetes target.
+#[derive(Clone)]
+pub struct Target {
+    selector: TargetSelector,
+    port: Port,
+    namespace: NameSpace,
+}
+
+/// A kubernetes namespace.
+/// If None, the default is "default".
+#[derive(Clone)]
+pub(crate) struct NameSpace(Option<String>);
+impl NameSpace {
+    /// Returns the configured namespace or the default.
+    pub(crate) fn name_any(&self) -> String {
+        let default = "default".to_string();
```

Comment:
Todo later: The default namespace should ideally be picked up from the kubeconfig or the flags(`--namespace`)..

Reply:
This is just the default for the type, the kubectl crate passes the namespace down as a parameter, so it should work like that already